File: Models/Flag.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using System.ComponentModel.DataAnnotations;

namespace BrewMapAPI.Models
{
    public class Flag
    {
        [BsonId]
        [BsonRepresentation(BsonType.ObjectId)]
        public string Id { get; set; } = String.Empty;

        [BsonElement("reportedByUserId")]
        [BsonRepresentation(BsonType.ObjectId)]
        public string ReportedByUserId { get; set; } = String.Empty;

        [BsonElement("target")]
        public ReportTarget Target { get; set; } = new ReportTarget();

        [BsonElement("reason")]
        public string Reason { get; set; } = String.Empty;

        [BsonElement("description")]
        public string? Description { get; set; }

        [BsonElement("status")]
        public string Status { get; set; } = "pending";

        [BsonElement("resolvedByAdminId")]
        [BsonRepresentation(BsonType.ObjectId)]
        public string? ResolvedByAdminId { get; set; }

        [BsonElement("resolvedAt")]
        public DateTime? ResolvedAt { get; set; }

        [BsonElement("resolutionNote")]
        public string? ResolutionNote { get; set; }

        [BsonElement("createdAt")]
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        [BsonElement("updatedAt")]
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    }

    public class ReportTarget
    {
        [BsonElement("type")]
        [AllowedValues("location", "product", "review", "user")]
        public string Type { get; set; } = String.Empty;

        [BsonElement("id")]
        [BsonRepresentation(BsonType.ObjectId)]
        public string TargetId { get; set; } = String.Empty;
    }
}

File: WebAPI/MobileApp/app/src/main/java/hr/algebra/mobileapp/models/flag/Flag.kt
package hr.algebra.mobileapp.models.flag

/**
 * Identifies the reported entity.
 *
 * [type] — one of: `"locationService"`, `"product"` (drinkService), `"reviewService"`, `"userService"`.
 * [id]   — MongoDB ObjectId string of the entity.
 */
data class ReportTarget(
    val type: String,
    val id: String
)

/** Request body for `POST api/Flag`. */
data class CreateFlagRequest(
    val target: ReportTarget,
    val reason: String,
    val description: String?
)

/**
 * Mirrors `ReadFlag` from the BrewMap API.
 *
 * [status] — one of: `"pending"`, `"reviewed"`, `"resolved"`.
 */
data class Flag(
    val id: String,
    val reportedByUserId: String,
    val target: ReportTarget,
    val reason: String,
    val description: String?,
    val status: String,
    val resolvedByAdminId: String?,
    val resolvedAt: String?,
    val resolutionNote: String?,
    val createdAt: String,
    val updatedAt: String
)
